File: src/main/python/AoC2023_25.py
#! /usr/bin/env python3
#
# Advent of Code 2023 Day 25
#

from __future__ import annotations

import random
import sys
from collections import defaultdict
from copy import deepcopy
from typing import NamedTuple

from aoc.common import InputData
from aoc.common import SolutionBase
from aoc.common import aoc_samples
from aoc.graph import flood_fill

TEST = """\
jqt: rhn xhk nvd
rsh: frs pzl lsr
xhk: hfx
cmg: qnr nvd lhk bvb
rhn: xhk bvb hfx
bvb: xhk hfx
pzl: lsr hfx nvd
qnr: nvd
ntq: jqt hfx bvb xhk
nvd: lhk
lsr: lhk
rzs: qnr cmg lsr rsh
frs: qnr lhk lsr
"""


class Graph(NamedTuple):
    edges: dict[str, list[str]]

    @classmethod
    def from_input(cls, input: InputData) -> Graph:
        edges = defaultdict[str, list[str]](list)
        for line in input:
            key, values = line.split(": ")
            edges[key] += [_ for _ in values.split()]
            for v in values.split():
                edges[v].append(key)
        return Graph(edges)

    def combine_nodes(self, node_1: str, node_2: str, new_node: str) -> Graph:
        self.edges[new_node] = [
            d for d in self.edges[node_1] if d != node_2
        ] + [d for d in self.edges[node_2] if d != node_1]
        for node in {node_1, node_2}:
            for dst in self.edges[node]:
                self.edges[dst] = [
                    new_node if d == node else d for d in self.edges[dst]
                ]
            del self.edges[node]
        return Graph(self.edges)

    def get_connected(self, node: str) -> set[str]:
        return flood_fill(node, lambda n: (_ for _ in self.edges[n]))


Input = Graph
Output1 = int
Output2 = str


class Solution(SolutionBase[Input, Output1, Output2]):
    def parse_input(self, input_data: InputData) -> Input:
        return Graph.from_input(input_data)

    def part_1(self, graph: Input) -> Output1:
        while True:
            g = deepcopy(graph)
            counts = {node: 1 for node in g.edges.keys()}
            while len(g.edges) > 2:
                a = random.sample(list(g.edges.keys()), 1)[0]
                b = random.sample(list(g.edges[a]), 1)[0]
                new_node = f"{a}-{b}"
                counts[new_node] = counts.get(a, 0) + counts.get(b, 0)
                del counts[a]
                del counts[b]
                g = g.combine_nodes(a, b, new_node)
            a, b = g.edges.keys()
            if len(g.edges[a]) == 3:
                return counts[a] * counts[b]

    def part_2(self, input: Input) -> Output2:
        return "🎄"

    @aoc_samples((("part_1", TEST, 54),))
    def samples(self) -> None:
        pass


solution = Solution(2023, 25)


def main() -> None:
    solution.run(sys.argv)


if __name__ == "__main__":
    main()

File: src/main/java/AoC2023_25.java
import static com.github.pareronia.aoc.Utils.concatAll;
import static java.util.stream.Collectors.toMap;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Random;
import java.util.Set;

import com.github.pareronia.aoc.StringOps;
import com.github.pareronia.aoc.StringOps.StringSplit;
import com.github.pareronia.aoc.solution.Sample;
import com.github.pareronia.aoc.solution.Samples;
import com.github.pareronia.aoc.solution.SolutionBase;

public final class AoC2023_25
        extends SolutionBase<AoC2023_25.Graph, Integer, String> {
    
    private AoC2023_25(final boolean debug) {
        super(debug);
    }
    
    public static AoC2023_25 create() {
        return new AoC2023_25(false);
    }
    
    public static AoC2023_25 createDebug() {
        return new AoC2023_25(true);
    }
    
    @Override
    protected Graph parseInput(final List<String> inputs) {
        return Graph.fromInput(inputs);
    }

    @Override
    public Integer solvePart1(final Graph graph) {
        final Random rand = new Random();
        while (true) {
            Graph g = graph.copy();
            final Map<String, Integer> counts = g.edges.keySet().stream()
                    .collect(toMap(k -> k, k -> 1));
            while (g.edges.size() > 2) {
                final String a = g.edges.keySet().stream()
                        .skip(rand.nextInt(g.edges.size() - 1))
                        .findFirst().orElseThrow();
                final String b = g.edges.get(a).get(rand.nextInt(g.edges.get(a).size()));
                final String newNode = "%s-%s".formatted(a, b);
                counts.put(newNode, counts.get(a) + counts.get(b));
                counts.remove(a);
                counts.remove(b);
                g = g.combineNodes(a, b, newNode);
            }
            final List<String> nodes = List.copyOf(g.edges.keySet());
            if (g.edges.get(nodes.get(0)).size() == 3) {
                return counts.get(nodes.get(0)) * counts.get(nodes.get(1));
            }
        }
    }
    
    @Override
    public String solvePart2(final Graph graph) {
        return "🎄";
    }
    
    @Samples({
        @Sample(method = "part1", input = TEST, expected = "54"),
    })
    public static void main(final String[] args) throws Exception {
        AoC2023_25.create().run();
    }

    private static final String TEST = """
            jqt: rhn xhk nvd
            rsh: frs pzl lsr
            xhk: hfx
            cmg: qnr nvd lhk bvb
            rhn: xhk bvb hfx
            bvb: xhk hfx
            pzl: lsr hfx nvd
            qnr: nvd
            ntq: jqt hfx bvb xhk
            nvd: lhk
            lsr: lhk
            rzs: qnr cmg lsr rsh
            frs: qnr lhk lsr
            """;
    
    record Graph(Map<String, List<String>> edges) {
        
        public static Graph fromInput(final List<String> input) {
            final HashMap<String, List<String>> edges = new HashMap<>();
            for (final String line : input) {
                final StringSplit splits = StringOps.splitOnce(line, ": ");
                final List<String> lst
                    = new ArrayList<>(Arrays.asList(splits.right().split(" ")));
                edges.computeIfAbsent(splits.left(), s -> new ArrayList<>()).addAll(lst);
                lst.forEach(dst ->
                    edges.computeIfAbsent(dst, d -> new ArrayList<String>())
                        .add(splits.left()));
            }
            return new Graph(edges);
        }
        
        public Graph copy() {
            return new Graph(this.edges.keySet().stream()
                .collect(toMap(
                        k -> k,
                        k -> this.edges.get(k).stream().toList())));
        }
        
        public Graph combineNodes(
                final String node1, final String node2, final String newNode
        ) {
            this.edges.put(newNode, concatAll(
                this.edges.get(node1).stream().filter(d -> !d.equals(node2)).toList(),
                this.edges.get(node2).stream().filter(d -> !d.equals(node1)).toList()
            ));
            for (final String node : Set.of(node1, node2)) {
                for (final String dst : this.edges.get(node)) {
                    final List<String> lst = this.edges.get(dst).stream()
                        .map(d -> d.equals(node) ? newNode : d)
                        .toList();
                    this.edges.put(dst, lst);
                }
                this.edges.remove(node);
            }
            return new Graph(this.edges);
        }
    }
}
